### PR TITLE
Allow EOF IO passed to JSON::PullParser.new

### PR DESCRIPTION
**Pull request by Blacksmoke16**
`EOF` is a valid pull parser token type.  Because of this, it makes sense to set the initial kind to `EOF` if that's the initial state of the input versus raising an exception.

Discovered via https://github.com/Blacksmoke16/oq/issues/82

### CODE REVIEW

**Review by straight-shoota — APPROVED** [review]

File: src/json/pull_parser.cr [hunk 1]
```diff
@@ -126,6 +126,8 @@ class JSON::PullParser
       begin_array
     when .begin_object?
       begin_object
+    when .eof?
```
Comment (asterite) [inline src/json/pull_parser.cr:129]: Can we change all of these `when` to be `in`? I wonder if other cases are missing as well (probably not, but using `in` is better anyways)

Comment (Blacksmoke16) [reply src/json/pull_parser.cr:129]:
```
Missing members:
 - EndArray
 - EndObject
 - Colon
 - Comma
```

I think it makes sense to raise on these given it means the input is invalid JSON.  If we really wanted we could make this `in` and have these 4 members included in a single `in` that raises?

Comment (asterite) [reply src/json/pull_parser.cr:129]: I think that works! 👍

**Comment by Blacksmoke16** [comment]
This is good for another review @asterite.

**Review by asterite — APPROVED** [review]
Looks good. On second thought, what's the use case for this? An empty json is invalid. I think json pull parser only produces valid json expressions, and eof is not one of them.

**Comment by Blacksmoke16** [comment]
@asterite The use case is for `oq` in the case where the `jq` filter doesn't output anything.  E.g. `jq '.[] | select(.name != "foo")' <<<$'[{"name":"foo"}]'`.  When using a filter like this with `oq`, and with the output format specified to something other than JSON, the pull parser is failing because the IO is already essentially empty, which is this line: https://github.com/Blacksmoke16/oq/blob/master/src/converters/yaml.cr#L11

```sh
oq -i yaml -o yaml '.[] | select(.name != "foo")' <<<$'- name: foo'
oq error: Unexpected token: <EOF> at line 1, column 1
```

The solution in `oq` land is of course to just return early if there is no data to process.  However I don't think there's a way to know that the related `IO` has no data, so this seemed like the most logical solution.  I suppose I could just rescue this exception but :shrug:.

**Comment by straight-shoota** [comment]
@asterite This change does not parse an empty document as valid. It just makes sure to properly forward the information that it has reached EOF on the input stream.

**Comment by asterite** [comment]
Thanks! Feel free to merge.

**Comment by straight-shoota** [comment]
I'm actually unsure if this qualifies as a bug-fix or should wait for the end of the feature freeze. 
(Thumbs up for bug fix)

**Comment by beta-ziliani** [comment]
I need an emoji for a horizontal thumb :-D really, both interpretations are valid here. I vote for merging it. BTW @Blacksmoke16 thanks for refactoring the case, that's also a nice improvement.

**Comment by straight-shoota** [comment]
I think I'd rather postpone this to 1.2. We're about to release 1.1 and there's an off chance that this might have any unforeseeable effect. It's not an urgent bug fix.

**Pull request merged**